### PR TITLE
Partial fix of tuefind issue 1043?

### PR DESCRIPTION
**Pull request by ruschein**

### CODE REVIEW

File: module/VuFind/src/VuFind/RecordDriver/DefaultRecord.php [hunk 1]
```diff
@@ -958,6 +958,24 @@ public function getPrimaryAuthor()
         return $authors[0] ?? '';
     }
 
+    static private $lang_code = $this->tuefind()->getTranslatorLocale();
```
Comment (mtrojan-ub) [inline module/VuFind/src/VuFind/RecordDriver/DefaultRecord.php:961]: You should never make any changes directly to the VuFind module. If they are IxTheo-related, they should be moved into the IxTheo RecordDriver. IxTheo already defines multiple functions which override the original VuFind author logic, so the changes need to be adjusted as well.

File: module/VuFind/src/VuFind/RecordDriver/DefaultRecord.php [hunk 2]
```diff
@@ -958,6 +958,24 @@ public function getPrimaryAuthor()
...
+
+    // Replaces any occurrences of "v.Chr." in $s w/ the version for the currently selected user-interface language.
+    static private function MapBeforeChrist(string $s): string {
+        global $lang_code;
```
Comment (mtrojan-ub) [inline module/VuFind/src/VuFind/RecordDriver/DefaultRecord.php:965]: lang_code is a class member, so there is no need to use the `global` keyword.

File: module/VuFind/src/VuFind/RecordDriver/DefaultRecord.php [hunk 3]
```diff
@@ -958,6 +958,24 @@ public function getPrimaryAuthor()
...
+        static $lang_code_to_bc_abbreviations_map = array(
+            "en" => "BC",
+            "de" => "v.Chr.",
+            "fr" => "avant J.-C.",
+            "it" => "a.C.",
+            "cn" => "公元前"
+        );
+        if (!array_key_exists($lang_code, $lang_code_to_bc_abbreviations_map))
```
Comment (mtrojan-ub) [inline module/VuFind/src/VuFind/RecordDriver/DefaultRecord.php:973]: access static lang_code by using self::lang_code.

**Comment by mtrojan-ub** [comment]
Also, VuFind uses camel case for variable names, so please do this as well.

File: module/VuFind/src/VuFind/RecordDriver/DefaultRecord.php [hunk 4]
```diff
@@ -958,6 +958,24 @@ public function getPrimaryAuthor()
         return $authors[0] ?? '';
     }
 
+    static private $lang_code = $this->tuefind()->getTranslatorLocale();
+
+    // Replaces any occurrences of "v.Chr." in $s w/ the version for the currently selected user-interface language.
+    static private function MapBeforeChrist(string $s): string {
+        global $lang_code;
+        static $lang_code_to_bc_abbreviations_map = array(
```
[the remark this answers is not in the document]
Comment (mtrojan-ub) [reply module/VuFind/src/VuFind/RecordDriver/DefaultRecord.php:966]: also, i think you cannot use the `static` keyword here similar to c++, so you should define this array as a class member similar to lang_code.

File: import/marc_tuefind.properties [hunk 5]
```diff
@@ -25,7 +25,7 @@ title = custom(de.unituebingen.ub.ubtools.solrmarcMixin.TuelibBiblioMixin), getM
 title_alt = 100t:130adfgklnpst:240a:246a:505t:700t:710t:711t:730adfgklnpst:740a:787t
 title_auth = custom(de.unituebingen.ub.ubtools.solrmarcMixin.TuelibBiblioMixin), getMainTitle
 title_short = 245a:249a, first
-title_sort = 245abkp, titleSortLower, custom_map(de.unituebingen.ub.ubtools.solrmarcMixin.TuelibBiblioMixin normalizeSortableString)
+title_sort = 245abkp, toLower, custom_map(de.unituebingen.ub.ubtools.solrmarcMixin.TuelibBiblioMixin normalizeSortableString)
```
[the remark this answers is not in the document]
Comment (ruschein) [reply import/marc_tuefind.properties:28]: No, we don't but, we already do a bunch of those things ourselves. In fact, what we strip is "[^\\p{Lu}\\p{Ll}\\p{Lt}\\p{Lo}\\p{N}]+" which covers all of this.

File: module/VuFind/src/VuFind/RecordDriver/DefaultRecord.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (ruschein) [reply module/VuFind/src/VuFind/RecordDriver/DefaultRecord.php:966]: According to the PHP documentation you can use static like this, too in PHP.

File: module/VuFind/src/VuFind/RecordDriver/DefaultRecord.php (hunk 2 above)
Comment (ruschein) [reply module/VuFind/src/VuFind/RecordDriver/DefaultRecord.php:965]: Changed.

File: module/VuFind/src/VuFind/RecordDriver/DefaultRecord.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (mtrojan-ub) [reply module/VuFind/src/VuFind/RecordDriver/DefaultRecord.php]: you're actually right, but anyway lang_code and lang_code_to_bc_abbreviations_map should be initialized the same way regarding their static behaviour.

File: module/VuFind/src/VuFind/RecordDriver/DefaultRecord.php (hunk 1 above)
Comment (mtrojan-ub) [reply module/VuFind/src/VuFind/RecordDriver/DefaultRecord.php]: If you moved this, your PR does actually not contain the IxTheo record driver at the moment.

Comment (ruschein) [reply module/VuFind/src/VuFind/RecordDriver/DefaultRecord.php]: It does now.

File: module/VuFind/src/VuFind/RecordDriver/DefaultRecord.php [hunk 6]
```diff
@@ -966,7 +966,7 @@ public function getPrimaryAuthor()
     public function getPrimaryAuthors()
     {
         return isset($this->fields['author'])
-            ? (array)$this->fields['author'] : [];
+            ? (array)MapBeforeChrist($this->fields['author']) : [];
```
[the remark this answers is not in the document]
Comment (ruschein) [reply module/VuFind/src/VuFind/RecordDriver/DefaultRecord.php]: Hopefully fixed now.

File: module/IxTheo/src/IxTheo/RecordDriver/SolrDefault.php [hunk 7]
```diff
@@ -152,6 +152,22 @@ public function getAuthorDataFields($index, $dataFields = [])
         return $data;
     }
 
+    // Replaces any occurrences of "v.Chr." in $s w/ the version for the currently selected user-interface language.
+    static private function MapBeforeChrist(string $s): string {
+        static $lang_code = this->tuefind()->getTranslatorLocale();
```
Comment (mtrojan-ub) [inline module/IxTheo/src/IxTheo/RecordDriver/SolrDefault.php:157]: this => $this

**Comment by mtrojan-ub** [comment]
I actually think that the MapBeforeChrist approach in the RecordDriver will not solve the Problem described in issue 1043. If this is really related to keyword translations, it might be necessary to apply a mapping to certain solr fields (e.g. era_facet_en, era_facet_de and so on). So maybe this normalization should be moved to Java.

@jriedl, can you please have a look at this PR too?

**Comment by ruschein** [comment]
> I actually think that the MapBeforeChrist approach in the RecordDriver will not solve the Problem described in issue 1043. If this is really related to keyword translations, it might be necessary to apply a mapping to certain solr fields (e.g. era_facet_en, era_facet_de and so on). So maybe this normalization should be moved to Java.
> 
> @jriedl, can you please have a look at this PR too?

While it may be possible to solve this problem w/ keyword translations, I suspect that it would greatly blow up the database.

**Comment by mtrojan-ub** [comment]
I think we should take a different approach for solving these issues, so i'm gonna close this one.

**Pull request closed without merging**